### DISCUSSION TITLE
Airflow Docker - ExternalPythonOperator - Python VENV locally on UBUNTU 20.04

### DISCUSSION THREAD

**Discussion by stromal** [apache/airflow#26783] [Q&A]
- I have a running official clean
- " https://airflow.apache.org/docs/apache-airflow/2.3.3/docker-compose.yaml " with python 3.8
- image: ${AIRFLOW_IMAGE_NAME:-apache/airflow:2.3.3-python3.8}
- My goal would be to have a few python virtual enironemnt locally  (that I also use to develope the DAG files and experiment with them) - I want to run the Airflow Docker continer's DAG files from some of these this local python virtual environemnts.
- For now it would ony be 1 python virtual environemnt with the same python 3.8
- I am not sure if it is possible as I am reading
   -  ""ExternalPythonOperator" - https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html#externalpythonoperator
   - https://github.com/apache/airflow/blob/main/LOCAL_VIRTUALENV.rst

**Comment by potiuk** [comment]
Yes. You need to go 2.4.0+ and use ExternalPythonOperator if you want those envs pre-baked into the image (You can extend the image and add those venvs while your are using your own env). You can also use PythonVirtualenvOperator to create those venvs dynamically and experiment with those. Or you can use DockerOperator/KubernetesOperator if you need isolation.

All the options are nicely summarized in this dedicated chapter - including discussion when to use which and how you can easily switch between development/iteration/experimentation mode  (PythonVirtualenv) and "deployment mode" (ExternalPythonOperator).

https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies

The 2.4.0 ExternalPythonOperator was specifically introduced in 2.4.0 to handle the very case you describe.

Note that this is NOT possible/advised to use Local host virtualenvs in Docker Compose - i.e. mounting host venv to docker container. If you want, you can try to do it but you are mostly on your own  - as it might have multiple problems and edge cases that you will have to fight with - for example if you have different version of shared c libraries in Host and Container, the compiled Python code (Rust, C etc - which is often used in data science, - pandas, numpy etc.) might have very strange and difficult to debug problems - sigsegvs etc are something to expec. Similarly if you have MacOS host and Linux container this might lead to even worse problems (not mentioning M1). So using virtualenv that's either pre-build  in ExternalPythonOperator (in the image of yours) or dynamically created using PythonVirtualenvOperator) is the only reasonable approach if you want to use Docker. 

You can still use multiple Host virtualenvs from your local virtualenv-installed Airflow (using ExternalPythonOperator).

Note that te virtualenvs are per task not per DAGs. You cannot (for now) parse your DAGs and execute whole dags in different virtualenv - you can execute individual Python* tasks in those. Separate runtime environment for "whole DAGs" will likely be implemented in 2.4 or 2.6 as result of  https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-46+Runtime+isolation+for+airflow+tasks+and+dag+parsing

**Reply by stromal** [reply]
@potiuk Just to make sure that I understand the approach that I have mentioned in my original description (Official Airflow Docker Image + outside of Docker Airflow python virtual env running on the same AWS EC2), you advise against that and I am on my own to do it?

- This make me see that I need to Update from 2.3.3 to 2.4.0 . 
- To do that I need a 2.4.0 Docker image that works.
- Is the dev version of the image is "log error" free now? https://github.com/apache/airflow/pull/26493 
- Because I am not sure how the fix should be implemented in the Docker version.
